<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <arg name="apriltag_settings" default="$(find tj2_bringup)/config/diffyjr/tags/settings.yaml"/>
    <arg name="apriltag_tags" default="$(find tj2_bringup)/config/diffyjr/tags/tags.yaml"/>
    <arg name="publish_tag_detections_image" default="false"/>

    <param name="use_sim_time" value="false" />

    <!-- <node pkg="robot_localization" type="ukf_localization_node" name="northstar_kalman_filter" clear_params="true">
        <rosparam command="load" file="$(find tj2_northstar)/config/ukf.yaml" />
    </node> -->

    <group ns="northstar">
        <node pkg="tj2_northstar" type="playback.py" name="arducam_playback" required="false" output="screen">
            <param name="info_directory" value="$(find tj2_bringup)/config/diffyjr/northstar/calibration" />
            <param name="bag_name" value="arducam_2023-05-15-16-30-52" />
        </node>

        <rosparam command="load" file="$(arg apriltag_settings)" ns="apriltag_cam_array"/>
        <rosparam command="load" file="$(arg apriltag_tags)" ns="apriltag_cam_array"/>
        <node pkg="tj2_northstar" type="apriltag_cam_array" name="apriltag_cam_array" required="false" output="screen">
            <remap from="info" to="camera/camera_info"/>
            <remap from="image_raw" to="camera/image_raw"/>
        </node>


        <node pkg="tj2_northstar" type="landmark_converter.py" name="landmark_converter" required="false" output="screen">
            <rosparam param="landmark_ids">[1, 2, 3, 4, 5, 6, 7, 8]</rosparam>
            <rosparam param="covariance">[
                0.0001, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
                0.0000, 0.0001, 0.0000, 0.0000, 0.0000, 0.0000,
                0.0000, 0.0000, 0.0010, 0.0000, 0.0000, 0.0000,
                0.0000, 0.0000, 0.0000, 0.0050, 0.0000, 0.0000,
                0.0000, 0.0000, 0.0000, 0.0000, 0.0050, 0.0000,
                0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0050,
            ]</rosparam>
        </node>

        <!-- <include file="$(find tj2_northstar)/launch/apriltag_single_cam.launch">
            <arg name="apriltag_settings" value="$(arg apriltag_settings)"/>
            <arg name="apriltag_tags" value="$(arg apriltag_tags)"/>
            <arg name="node_name" value="apriltag_camera_0"/>
            <arg name="camera_info_topic" value="camera_0/camera_info"/>
            <arg name="image_rect_topic" value="camera_0/image_raw"/>
        </include>
        <include file="$(find tj2_northstar)/launch/apriltag_single_cam.launch">
            <arg name="apriltag_settings" value="$(arg apriltag_settings)"/>
            <arg name="apriltag_tags" value="$(arg apriltag_tags)"/>
            <arg name="node_name" value="apriltag_camera_1"/>
            <arg name="camera_info_topic" value="camera_1/camera_info"/>
            <arg name="image_rect_topic" value="camera_1/image_raw"/>
        </include>
        <include file="$(find tj2_northstar)/launch/apriltag_single_cam.launch">
            <arg name="apriltag_settings" value="$(arg apriltag_settings)"/>
            <arg name="apriltag_tags" value="$(arg apriltag_tags)"/>
            <arg name="node_name" value="apriltag_camera_2"/>
            <arg name="camera_info_topic" value="camera_2/camera_info"/>
            <arg name="image_rect_topic" value="camera_2/image_raw"/>
        </include>
        <include file="$(find tj2_northstar)/launch/apriltag_single_cam.launch">
            <arg name="apriltag_settings" value="$(arg apriltag_settings)"/>
            <arg name="apriltag_tags" value="$(arg apriltag_tags)"/>
            <arg name="node_name" value="apriltag_camera_3"/>
            <arg name="camera_info_topic" value="camera_3/camera_info"/>
            <arg name="image_rect_topic" value="camera_3/image_raw"/>
        </include> -->

        <node pkg="tj2_camera" type="tag_marker_publisher.py" name="tag_marker_publisher" required="false" output="screen">
            <param name="marker_publish_rate" value="50"/>
        </node>
    </group>

    <node pkg="tf2_ros" type="static_transform_publisher" name="base_link_to_footprint" args="0.0 0.0 0.0  0 0 0 1 base_link base_tilt_link" />

    <node pkg="tf2_ros" type="static_transform_publisher" name="base_to_siren_center" args="0.1577 0.0 0.2867  0 0 0 1 base_tilt_link siren_center" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="siren_center_to_pivot_0" args="0.0 0.0 0.0  0.9969  -0.0000  0.0785  0.0000 siren_center pivot_0" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="siren_center_to_pivot_1" args="0.0 0.0 0.0  0.7049  0.7049  0.0555  -0.0555 siren_center pivot_1" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="siren_center_to_pivot_2" args="0.0 0.0 0.0  0.0000  0.9969  0.0000  -0.0785 siren_center pivot_2" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="siren_center_to_pivot_3" args="0.0 0.0 0.0  -0.7049  0.7049  -0.0555  -0.0555 siren_center pivot_3" />

    <node pkg="tf2_ros" type="static_transform_publisher" name="pivot_to_camera_0" args="0.0411 0.0 0.0033  0 0 0 1 pivot_0 camera_0" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="pivot_to_camera_1" args="0.0411 0.0 0.0033  0 0 0 1 pivot_1 camera_1" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="pivot_to_camera_2" args="0.0411 0.0 0.0033  0 0 0 1 pivot_2 camera_2" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="pivot_to_camera_3" args="0.0411 0.0 0.0033  0 0 0 1 pivot_3 camera_3" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="camera_to_optical_0" args="0.0 0.0 0.0  -0.5000  0.5000  -0.5000  0.5000 camera_0 camera_optical_0" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="camera_to_optical_1" args="0.0 0.0 0.0  -0.5000  0.5000  -0.5000  0.5000 camera_1 camera_optical_1" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="camera_to_optical_2" args="0.0 0.0 0.0  -0.5000  0.5000  -0.5000  0.5000 camera_2 camera_optical_2" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="camera_to_optical_3" args="0.0 0.0 0.0  -0.5000  0.5000  -0.5000  0.5000 camera_3 camera_optical_3" />

</launch>
